Remotion/Web/Development.WebTesting/Utilities/ProcessUtils.cs

```diff
@@ -160,14 +211,40 @@ public static void GracefulProcessShutdown ([NotNull] Process process, TimeSpan
         // https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.process.kill#System_Diagnostics_Process_Kill
         // This exception can be swallowed safely, as we are sure that a process is associated and the process has
         // already exited, therefore the method can return at this point.
-        return;
+      }
+    }
+
+    /// <summary>
+    /// Waits for the specified <paramref name="processes"/> to exit or until the <paramref name="sharedTimeout"/> is reached.
+    /// </summary>
+    /// <param name="processes">The processes that should be waited on to exit.</param>
+    /// <param name="sharedTimeout">The wait timeout after which waiting will be canceled.</param>
+    /// <returns>A subset of the specified <paramref name="processes"/> that did not exit withing the <paramref name="sharedTimeout"/>.</returns>
+    [MustUseReturnValue]
+    private static IReadOnlyList<Process> WaitForProcessesExits (IEnumerable<Process> processes, TimeSpan sharedTimeout)
+    {
+      var remainingProcesses = new List<Process>();
+
+      // Go through all processes and call .WaitForExit on each using a shared timeout
+      // If the timeout runs out still go through all remaining process to make sure to clear any exited processes
+      var stopwatch = Stopwatch.StartNew();
+      foreach (var process in processes)
+      {
+        var timeToTimeoutEnd = sharedTimeout - stopwatch.Elapsed;
+
+        // Use 0 (process.WaitForExit returns instantly) if we exceeded our timeout, otherwise use the remaining timeout
+        var remainingTimeoutInMilliseconds = Math.Max ((int) timeToTimeoutEnd.TotalMilliseconds, 0);
+        if (!process.WaitForExit (remainingTimeoutInMilliseconds))
+          remainingProcesses.Add (process);
       }
 
-      // No exception is thrown when calling .WaitForExit(..) on an already exited process.
-      if (process.WaitForExit (timeoutInMilliseconds))
-        return;
+      return remainingProcesses;
+    }
 
-      throw new InvalidOperationException (string.Format ("Process '{0}' (id: '{1}') did not exit in the expected amount of time.", process.ProcessName, process.Id));
+    private static Exception CreateProcessDidNotExitInTimeException (Process process)
+    {
+      return new InvalidOperationException (
```

Comment:
1. pass the timeout and use it in the message
2. add all processes, that way we can forgoe the aggregateexception
The following processes did not exit within the timeout of {0} seconds:
- Process 'Name' (id: ID)
- ...

ID is an INT, don't need quotes when writing {id: 123)